### ISSUE TITLE
[RFC] Make TorchFT checkpoint healing persistent and FSDP-safe

### ISSUE THREAD

**Issue by Skuld7451** [pytorch/torchtitan#4209]
# RFC 0004: Persistent checkpoint and FSDP-safe peer healing

## Summary

Make TorchTitan register model and optimizer state for TorchFT peer healing even
when persistent checkpoint saving is disabled, persist TorchFT manager state in
full checkpoints, establish one full-checkpoint writer, and isolate
replica-specific dataloader state from the shared checkpoint.

FSDP state materialization is explicitly requested on the training thread using
TorchFT's selective state-dict API.

## Dependencies

- Requires meta-pytorch/torchft#347 for the selective
  `state_dict_on_training_thread` API, first-quorum ordering, and safe
  state-read ordering around optimizer and device work.
- Does not require the TorchTitan process-group registry, synchronous
  integration, or LocalSGD CPU-staging changes.

## Problem

Persistent DCP checkpointing and in-memory TorchFT peer healing are different
capabilities. The previous integration returned early when persistent saving
was disabled, which also prevented registration of model and optimizer state
needed for peer healing.

There were additional ownership and state-layout problems:

- Multiple replicas could write the same full checkpoint path.
- Before the first quorum, no dynamic participating rank existed to select a
  writer.
- Dataloader state is replica-specific and cannot safely be loaded as one
  shared full-checkpoint state after membership changes.
- Materializing an FSDP state dict on TorchFT's recovery thread can enter
  replica-local collectives out of phase with the training ranks.
- A resumed full checkpoint did not carry TorchFT manager committed-step state.

### Healing state is independent of persistent saving

`torchtitan/experiments/torchft/checkpoint.py:88-152` retains the caller's input
states and constructs a healing view whenever TorchFT is enabled. If persistent
saving is disabled, model, optimizer, and scheduler wrappers are still created:

```python
if self.ft_manager:
    optimizers.init_cache_state_dict()

    if self.enable:
        self.states[TORCHFT_MANAGER] = self.ft_manager
        healing_states = self.states
    else:
        healing_states = {
            **input_states,
            MODEL: ModelWrapper(model_parts),
            OPTIMIZER: optimizers,
            LR_SCHEDULER: lr_schedulers,
        }
```

The callback includes model, optimizer, scheduler, and trainer state. Persistent
checkpoint `save()` remains a no-op when `checkpoint.enable=False`.

### FSDP state is materialized on the training thread

At `checkpoint.py:142-147`, TorchTitan opts this callback into TorchFT's
training-thread policy:

```python
self.ft_manager.register_state_dict_fn(
    "torchtitan",
    load_state_dict,
    state_dict,
    state_dict_on_training_thread=True,
)
```

This prevents TorchFT's recovery thread from independently entering the
replica-local collectives used by FSDP state-dict materialization. It directly
depends on a TorchFT release that provides the additive
`state_dict_on_training_thread` argument and first-quorum ordering described by
the FSDP-safe recovery API.

### One owner writes the shared full checkpoint

`checkpoint.py:201-209` selects the writer:

```python
def _owns_full_checkpoint(self) -> bool:
    if self.ft_manager is None:
        return True

    participating_rank = self.ft_manager.participating_rank()
    if participating_rank is not None:
        return participating_rank == 0

    return self.ft_replica_id == 0
```

After a quorum exists, participating rank 0 owns the shared full checkpoint.
Before the first quorum, configured replica 0 is the deterministic owner.
`_save()` at lines 163-178 returns the base manager's protected save result for
the owner and returns `False` for every other replica.

The public `save()` lifecycle, disabled guard, interval checks, and asynchronous
scheduling remain owned by `BaseCheckpointManager`. The TorchFT subclass only
implements the protected hooks expected by the current base class. `_load()` at
lines 180-184 first restores the replica-private dataloader and then delegates
the shared checkpoint load. `_wait_for_saving()` at lines 211-221 also handles
the asynchronous future created by the private dataloader save when the full
checkpoint async mode itself is disabled.

### Persistent manager state and dataloader isolation

`TORCHFT_MANAGER = "torchft_manager"` at `checkpoint.py:43` adds manager state to
the full DCP state so committed-step metadata survives resume.

`checkpoint.py:186-199` removes dataloader state from both full-checkpoint load
and flattened save paths whenever TorchFT is active:

```python
def _states_to_load(self, model_only: bool) -> dict[str, Any]:
    states = super()._states_to_load(model_only)
    if self.ft_manager:
        states.pop(DATALOADER, None)
    return states
```

When `enable_ft_dataloader_checkpoints=True`, each replica instead stores its
dataloader state under its own `ft-replicat-<id>` directory. When disabled,
exact dataloader recovery is intentionally unavailable and samples may replay.
This limitation is documented at
`torchtitan/experiments/torchft/README.md:83-99`.

## Compatibility and dependencies

- Jobs without TorchFT retain the base checkpoint manager behavior.
- Public checkpoint lifecycle and disabled-save behavior remain controlled by
  `BaseCheckpointManager`; the integration overrides only its protected hooks.
- Turning persistent checkpoint saving off no longer disables in-memory peer
  healing.
- Full-checkpoint ownership changes only multi-replica TorchFT jobs and prevents
  concurrent writes to the same DCP path.
- Dataloader state is no longer part of a shared TorchFT full checkpoint;
  replica-local saving must be enabled for exact data-position recovery.
- This integration directly requires the TorchFT selective training-thread
  state-dict API. It does not depend on the process-group registry, LocalSGD CPU
  staging, or synchronous integration changes.

## Validation

`torchtitan/experiments/torchft/tests/test_torchft_checkpoint.py:162-284` adds
coverage for healing with persistent checkpoints disabled, disabled-save no-op,
manager-state persistence, dataloader exclusion, dynamic owner selection, and
the replica-0 startup fallback. The tests mock `CheckpointManager._save`, which
verifies owner delegation without bypassing the public base lifecycle. The
existing asynchronous dataloader checkpoint test remains enabled and exercises
the `_wait_for_saving()` override.

The test also asserts that TorchTitan passes
`state_dict_on_training_thread=True`, so an older incompatible TorchFT API fails
immediately instead of silently using the unsafe execution thread.

Remote combined validation against the TorchFT FSDP-safe recovery branch
passed:

```text
pytest torchtitan/experiments/torchft/tests/test_torchft_checkpoint.py
7 passed
```
